### ISSUE TITLE
extra-include-dirs breaks use of precompiled packages in custom snapshots

### ISSUE THREAD

**Issue by liskin** [commercialhaskell/stack#3192]
### Steps to reproduce

1. `lts-7-x.yaml`:
    ```
    resolver: lts-7.14
    packages: [directory-1.3.1.1]
    ```
    `stack.yaml`:
    ```
    resolver:
      name: lts-7-x
      location: lts-7-x.yaml
    packages: []
    extra-include-dirs: ['/tmp/xxx']
    ```
1. `stack ghci --package directory`

### Actual

```
$ stack ghci --package directory
directory-1.3.1.1: configure
directory-1.3.1.1: build
directory-1.3.1.1: copy/register
```

### Expected

```
$ stack ghci --package directory
directory-1.3.1.1: using precompiled package
```

Which is exactly what happens when I set `extra-include-dirs` to `[]`. Also, if I run stack with `extra-include-dirs: []`, it populates the custom snapshot pkgdb with precompiled packages and uses them later when I add `extra-include-dirs` back, meaning its value is **not** used in the snapshot hash.

### Stack version

```
$ stack --version
Version 1.4.0, Git revision e714f1dd3fade19496d91bd6a017e435a96a6bcd (4640 commits) i386 hpack-0.17.0
```

### Method of installation

* Official binary, downloaded from stackage.org or fpcomplete's package repository

**Comment by liskin** [comment]
The reproducer is admittedly somewhat stupid and assumes that there's another snapshot somewhere which already has directory-1.3.1.1. Sorry about that.

**Comment by liskin** [comment]
So here's a much better and deterministic reproducer:

`lts-7-x.yaml`:
```
resolver: lts-7.14
packages: [directory-1.3.1.1]
```

`lts-7-y.yaml`:
```
resolver: lts-7.14
packages: [directory-1.3.1.1]
```

`stack1.yaml`:
```
resolver:
  name: lts-7-x
  location: lts-7-x.yaml
packages: []
extra-include-dirs: []
```

`stack2.yaml`:
```
resolver:
  name: lts-7-y
  location: lts-7-y.yaml
packages: []
extra-include-dirs: []
```

`stack3.yaml`:
```
resolver:
  name: lts-7-y
  location: lts-7-y.yaml
packages: []
extra-include-dirs: ['/tmp/xxx']
```

Human readable summary:
`lts-7-x.yaml` and `lts-7-y.yaml` are exactly the same, just different snapshot names.
`stack1.yaml` and `stack2.yaml` differ in `resolver` only.
`stack2.yaml` and `stack3.yaml` differ in `extra-include-dirs` only.

```
$ stack --system-ghc --stack-root "$(pwd)"/sr1 --stack-yaml stack1.yaml update
…
$ stack --system-ghc --stack-root "$(pwd)"/sr1 --stack-yaml stack1.yaml build directory
…
directory-1.3.1.1: download
directory-1.3.1.1: configure
directory-1.3.1.1: build
directory-1.3.1.1: copy/register
$ stack --system-ghc --stack-root "$(pwd)"/sr1 --stack-yaml stack2.yaml build directory
directory-1.3.1.1: using precompiled package
$ stack --system-ghc --stack-root "$(pwd)"/sr1 --stack-yaml stack3.yaml build directory
$ 
```

```
$ stack --system-ghc --stack-root "$(pwd)"/sr2 --stack-yaml stack1.yaml update
…
$ stack --system-ghc --stack-root "$(pwd)"/sr2 --stack-yaml stack1.yaml build directory
…
directory-1.3.1.1: download
directory-1.3.1.1: configure
directory-1.3.1.1: build
directory-1.3.1.1: copy/register
$ stack --system-ghc --stack-root "$(pwd)"/sr2 --stack-yaml stack3.yaml build directory
directory-1.3.1.1: configure
directory-1.3.1.1: build
directory-1.3.1.1: copy/register
$ stack --system-ghc --stack-root "$(pwd)"/sr2 --stack-yaml stack2.yaml build directory
$ 
```

I'd expect both scenarios to build the package just once and then use a precompiled package.

**Comment by liskin** [comment]
Another observation is that in the second scenario, `--extra-include-dirs=/tmp/xxx` is passed to setup of the `directory` package so it kind of makes sense that the hash used for precompiled lookup is different. It is suprising, however, that this happens even with `apply-ghc-options: locals` as it clearly breaks invariants about snapshot database just like `apply-ghc-options: everything` does.

If there was a way to specify `extra-include-dirs` for local packages only, we'd use that and this issue could be considered resolved. Ideally there'd be something like `setup-options` with the same syntax as `ghc-options`/`flags` so that we can specify extra include dirs only for packages that actually need them.

**Comment by mpilgrem** [comment]
@liskin, I am working through issues - including old ones. You were experiencing this with Stack 1.4.1. Stack 1.6.1 made some major changes, as did Stack 2.1.1. Do you know if what you were experiencing still applies?

**Comment by liskin** [comment]
@mpilgrem I don't think I remember much about this issue, so we are both in the same starting position if we are to follow the reproducer and try to figure out if it's still an issue. That being said, I left the day job where I encountered this issue years ago, and I have upgraded my laptop since, so to be entirely honest, I really don't care at all whether this is closed or resolved or forgotten or whatever. After all, if anyone ever runs into this in the future, it can always be reopened.
